### PR TITLE
Automate publishing Python package on PyPI

### PR DESCRIPTION
**Pull request by bittner**
Adds two Tox environments that help with publishing the Python package on PyPI.

- `tox -e package` can be used for linting to help detect packaging issues.
- `tox -e package -- upload` allows to publish the package on PyPI.
- `tox -e ensure_version_matches $GIT_TAG` verifies that the package version matches the semantic Git tag on the current commit. This is used by the GHA configuration to prevent a mismatch of tags and the version of the published package.

To make publishing work, you need to add the 2 secrets `PYPI_USERNAME` and `PYPI_PASSWORD` to [Settings ➜ Secrets ➜ Actions](https://github.com/canonical/charmcraft/settings/secrets/actions) with the PyPI access token for the project, as described in the [Python packaging user guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/#uploading-the-distribution-archives).

## Related

- #1579
- [Is `setup.py` deprecated?](https://packaging.python.org/en/latest/discussions/setup-py-deprecated/) – _The use as a command line tool is deprecated_

### CODE REVIEW

File: .github/workflows/publish-pypi.yaml
Comment (lengau) [inline .github/workflows/publish-pypi.yaml:6]: Could we make this an `x.y.z` tag format instead? We sometimes use tags for other stuff (e.g. marking a specific snap revision) and I don't want that to trigger the workflow even if it won't publish.

**Review by lengau — COMMENTED** [review]
Thanks for this!

I *think* snapcraft 8 lets us get rid of the `setup.py` file, but we'll find out when I run the CI

File: .github/workflows/publish-pypi.yaml
Comment (bittner) [reply .github/workflows/publish-pypi.yaml]: Good point. I modified the selector following GHA's [filter pattern cheat sheet](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet).

Comment (lengau) [reply .github/workflows/publish-pypi.yaml]: Thanks!

**Comment by bittner** [comment]
> I _think_ snapcraft 8 lets us get rid of the `setup.py` file, but we'll find out when I run the CI

The `snap-build` job fails, because the charmcraft snap configuration takes for granted that `setup.py` is around:

```
:: ++ python3 setup.py --version
:: python3: can't open file '/root/parts/charmcraft/src/setup.py': [Errno 2] No such file or directory
```

From a Python (packaging) point of view the use of `setup.py` is legacy and every modern Python package should switch over to using (just) `pyproject.toml`. I guess it's okay if I try to fix the snap configuration with an additional commit to make the pipeline pass.

**Comment by bittner** [comment]
The `snap-build` job fails as it can't find the `craft_cli` module.

```python
ModuleNotFoundError: No module named 'craft_cli'
```

I can't see that the `craft-cli` Python package is installed in the job, I can only see it as a dependency in [requirements.txt](https://github.com/canonical/charmcraft/blob/main/requirements.txt#L7).

Can anyone see what is the problem?

**Comment by bittner** [comment]
> Can anyone see what is the problem?

Okay, looks like there is a problem with the `sed` command before that:

```
sed: -e expression #1, char 14: unterminated `s' command
```

The problem is that that with the new setup the returned version string also contains the package name. I'll try to find a solution. Suggestions welcome!

**Comment by bittner** [comment]
The CI jobs all pass now (on my personal branch). I rebased the branch, and I made separate commits to document the specific changes that were necessary to work around a few build issues. :scroll: 

Hope that looks good enough now for the PR to be merged. :crossed_fingers:

**Comment by bittner** [comment]
I have signed the [CLA](https://ubuntu.com/legal/contributors) now. I use a different email address on Launchpad than on GitHub, that's why the check probably didn't find my Launchpad account.

**Comment by bittner** [comment]
The cla-check still fails. I checked my Launchpad account again; looks like there was another account with the email address I use on GitHub. I merged the accounts on Launchpad now.

**Comment by bittner** [comment]
I received an email notification this morning saying, I was added as a member of [Canonical Contributor Agreement](https://launchpad.net/~contributor-agreement-canonical).

I assume, the cla-check job would now succeed.

File: pyproject.toml
Comment (lengau) [inline pyproject.toml:7]: What drove this need? Please include a comment so we know when we can undo it

File: snap/snapcraft.yaml
Comment (lengau) [inline snap/snapcraft.yaml:141]: This change seems to be always making the charmcraft snap that gets built report the version `dev`, I'm guessing because it's not recognised as installed at this point, so there's not a `_version.py` file in the charmcraft directory from which to report.

**Review by lengau — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (bittner) [reply pyproject.toml:7]: The commit message says it all.

> Stick to older attr version to avoid hatchling UnknownPluginError (fancy-pypi-readme)

File: snap/snapcraft.yaml
Comment (bittner) [reply snap/snapcraft.yaml:141]: Doesn't setuptools-scm take care of generating the `_version.py` file, transparently? Have you seen any issues in the build process? :astonished:

**Comment by bittner** [comment]
@lengau Let me know what needs to be changed, if anything, so that I can help getting this merged. Thank you! :+1:

**Comment by bittner** [comment]
Is there a reason why this PR isn't considered for merging? Any arguments against the changes? Anything that needs to be updated?

It would be very helpful if the recent release(s) of charmcraft were published also on PyPI. For example, I'm currently blocked by the [<q>No keyring found</q> error](https://discourse.charmhub.io/t/charmcraft-and-the-no-keyring-found-error/6144) that was [fixed in version 2.3.0](https://github.com/canonical/charmcraft/releases/tag/2.3.0) by allowing anonymous downloads of charm library modules. As explained in https://github.com/canonical/charmcraft/issues/1579#issuecomment-2001979224 a pure-Python installation is needed for the CI use case.

File: snap/snapcraft.yaml
Comment (lengau) [reply snap/snapcraft.yaml:141]: The [snap artefact from the build](https://github.com/canonical/charmcraft/actions/runs/8500696494?pr=1580) is reporting that, and I've been unable to come up with something else that makes it work as intended. Could you revert the removal of `setup.py` for now, since having it there shouldn't break things?

File: requirements.txt
Comment (syu-w) [inline requirements.txt:1]: Version downgrade?

**Review by lengau — COMMENTED** [review]
Please re-add setup.py, and then I'm happy with this :-)

I've created https://github.com/canonical/charmcraft/issues/1635 for seeing if we can remove setup.py at a later date so we don't block this PR with it

File: snap/snapcraft.yaml
Comment (bittner) [reply snap/snapcraft.yaml]: `setup.py` restored. The built `snap.zip` bears a version number again.

**Comment by bittner** [comment]
***ping*** :bell: 

@lengau Alex, I think this PR is now ready for merging. Can you (or some other maintainer) take a look?

**Review by lengau — APPROVED** [review]

File: tox.ini
Comment (mattculler) [inline tox.ini:128]: I think it would be cleaner to do something like:
```
commands = python -c 'import os;' \
  'from importlib.metadata import version;' \
  'pkg, tag = os.environ["PKG_NAME"], os.environ["GIT_TAG"];' \
  ...
```
But if this doesn't work for some reason, or there's disagreement on this point, then I'm ok with it as-is.

**Review by mattculler — APPROVED** [review]
👍

File: tox.ini
Comment (lengau) [reply tox.ini:128]: This doesn't work in tox for some weird reasons :-)

**Comment by bittner** [comment]
Thanks for merging! 💯 

Now don't forget to add the secrets to the project settings to make automatic publishing succeed! 🤖

**Pull request merged**
